File: Top Interview Questions/Array/Two Sum/Two Sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        # 76ms,14.5MB
        d={}
        for i in range(len(nums)):
            x = target-nums[i]
            if x in d:
                return [d[x],i]

            d[nums[i]]=i

        return []
        
        
"""
# 56ms, 14.6MB
d={}

for i in range(len(nums)):
    fark = target-nums[i]
    if fark in d:
        return [d[fark],i]

    if nums[i] not in d:
        d[nums[i]]=i
"""

File: leetcode-interview-swift/leetcode-interview-swift/Easy/Array/two_sum.swift
//
//  two_sum.swift
//  leetcode-interview-swift
//
//  Created by Yusuf Turan on 27.02.2021.
//

//brute force
func twoSumBrute(_ nums: [Int], _ target: Int) -> [Int] {
	var sums = Array<Int>()
	for i in 0..<nums.count {
		for j in (i+1)..<nums.count {
			if (nums[i] + nums[j] == target) {
				sums.append(i)
				sums.append(j)
			}
		}
	}
	return sums
}

//hash:
func twoSum(_ nums: [Int], _ target: Int) -> [Int] {
	guard !nums.isEmpty else {
		return [-1, -1]
	}
	var indices = Array<Int>()
	var hashTable = Dictionary<Int, Int>()
	for (index, val) in nums.enumerated() {
		if let right = hashTable[val] {
			indices.append(index)
			indices.append(right)
			return indices
		}
		hashTable[target-val] = index
	}
	return indices
}
